salt/modules/win_event_viewer.py

```diff
@@ -0,0 +1,285 @@
+    '''
+
+    try:
+        return win32evtlog.OpenEventLog(target_computer, log_name)
+    except pywintypes.error:
+        raise FileNotFoundError('Log "{0}" of "{1}" can not be found or access was denied!'.format(log_name,
+                                                                                                   target_computer))
+
+
+def _close_event_handler(handler):
+    '''
+    info: will close the event handler
+    :param handler: PyHANDLE
+    :return:
+    '''
+
+    win32evtlog.CloseEventLog(handler)
+
+
+def get_event_generator(log_name, target_computer=None, raw=False):
+    '''
+    info: will get all log events one by one
+    warning: events are not in exact order
+    :param log_name: str
+    :param target_computer: None or str
+    :param raw: bool: True: PyEventLogRecord False: dict
+    :return: PyEventLogRecord or dict
+    '''
+
+    handler = _get_event_handler(log_name, target_computer)
+    flags = win32evtlog.EVENTLOG_BACKWARDS_READ | win32evtlog.EVENTLOG_SEQUENTIAL_READ
+    event_count = 0
+
+    while win32evtlog.GetNumberOfEventLogRecords(handler) > event_count:
+        # get list of some of the events
+        events = win32evtlog.ReadEventLog(handler, flags, 0)
+        if not events:
+            # event log was updated and events are not ready to be given yet
+            # rather than wait just return
+            break
+        for event in events:
+            event_count += 1
+            if raw:
+                yield event
+            else:
+                yield make_event_dict(event)
+    _close_event_handler(handler)
+
+
+def get_events(log_name, target_computer=None, raw=False):
+    '''
+    info: will make a tuple of log events
+    :param log_name: str
+    :param target_computer: None or str
+    :param raw: bool: True: PyEventLogRecord False: dict
+    :return: tuple
+    '''
+
+    return tuple(get_event_generator(log_name, target_computer, raw))
+
+
+def get_event_sorted_by_info_generator(log_name, target_computer=None):
+    '''
+    info: makes keys to event
+    :param log_name: str
+    :param target_computer: None or str
+    :return: dict
+    '''
+
+    for event in get_event_generator(log_name, target_computer):
+        event_info = {}
+        for part in event:
+            event_info[part] = event.get(part)
+
+        for key in TIME_PARTS:
+            event_info[key] = event.get('timeGenerated')[TIME_PARTS.get(key)]
+
+        yield event, event_info
+
+
+def get_events_sorted_by_info(log_name, target_computer=None):
+    '''
+    info: keys to list of events that match that key
+    :param log_name: str
+    :param target_computer: None or str
+    :return: dict
+    '''
+
+    event_info = {event_part: {} for event_part in EVENT_PARTS + tuple(TIME_PARTS.keys())}
+    for event, info in get_event_sorted_by_info_generator(log_name, target_computer):
+        for part in info:
+            event_info.get(part).setdefault(info.get(part), []).append(event)
```

Comment:
I'm sorting the events by their info and placing them in a dick like this 
{'eventID': {34: [e1, e4], 1: [e2], 4: [e3]},
 'sid': {None: [e1, e2, e3], 'somthing': [e4]}}

Reply:
I have not used collections.defaultdict(list)... Do you believe it can help me with make that dict?

Reply:
You'd do it like `event_info = {event_part: defaultdict(list) ...}` then use it with:

```
event_info[part].append(event)
```

Reply:
I see what you want me to do but I have a dict->dict->list not a dict->list and I cant just do dict->list because parts can have the same value

Reply:
@cmcmarrow perhaps this will explain what I mean better:

```
import pprint
from collections import defaultdict
PARTS = ['a', 'c', 'b', 'd', 'z', 'e']
INFO = (
    ('a', {'b': 'B', 'z': 'Z'}),
    ('c', {'d': 'D', 'e': 'E'}),
)


event_info = {event_part: {} for event_part in PARTS}
for event, info in INFO:
    for part in info:
        event_info.get(part).setdefault(info.get(part), []).append(event)
pprint.pprint(event_info)

event_info = {event_part: defaultdict(list) for event_part in PARTS}
for event, info in INFO:
    for part in info:
        event_info[part][info.get(part)].append(event)
pprint.pprint(event_info) 
```

Reply:
Okay I understand now. Thanks